cpp/src/arrow/array/validate.cc

```diff
@@ -392,96 +376,159 @@ Status ValidateArray(const Array& array) {
+  Status Validate() { return ValidateWithType(*data.type); }
 
-  Status Visit(const LargeBinaryArray& array) { return ValidateBinaryArray(array); }
+  Status ValidateWithType(const DataType& type) { return VisitTypeInline(type, this); }
 
-  Status Visit(const ListArray& array) { return ValidateListArray(array); }
+  Status Visit(const NullType& type) { return Status::OK(); }
 
-  Status Visit(const LargeListArray& array) { return ValidateListArray(array); }
+  Status Visit(const FixedWidthType& type) { return Status::OK(); }
 
-  Status Visit(const MapArray& array) {
-    // TODO check keys and items individually?
-    return ValidateListArray(array);
+  Status Visit(const StringType& type) {
+    RETURN_NOT_OK(ValidateBinaryLike(type));
+    return ValidateUTF8(data);
   }
 
-  Status Visit(const UnionArray& array) {
-    const auto& child_ids = array.union_type()->child_ids();
+  Status Visit(const LargeStringType& type) {
+    RETURN_NOT_OK(ValidateBinaryLike(type));
+    return ValidateUTF8(data);
+  }
 
-    const int8_t* type_codes = array.raw_type_codes();
-    for (int64_t i = 0; i < array.length(); ++i) {
-      if (array.IsNull(i)) {
-        continue;
+  Status Visit(const BinaryType& type) { return ValidateBinaryLike(type); }
+
+  Status Visit(const LargeBinaryType& type) { return ValidateBinaryLike(type); }
+
+  Status Visit(const ListType& type) { return ValidateListLike(type); }
+
+  Status Visit(const LargeListType& type) { return ValidateListLike(type); }
+
+  Status Visit(const MapType& type) { return ValidateListLike(type); }
+
+  Status Visit(const FixedSizeListType& type) {
+    const ArrayData& child = *data.child_data[0];
+    const Status child_valid = ValidateArrayFull(child);
+    if (!child_valid.ok()) {
+      return Status::Invalid("Fixed size list child array invalid: ",
+                             child_valid.ToString());
+    }
+    return Status::OK();
+  }
+
+  Status Visit(const StructType& type) {
+    // Validate children
+    for (int64_t i = 0; i < type.num_fields(); ++i) {
+      const ArrayData& field = *data.child_data[i];
+      const Status field_valid = ValidateArrayFull(field);
+      if (!field_valid.ok()) {
+        return Status::Invalid("Struct child array #", i,
+                               " invalid: ", field_valid.ToString());
       }
+    }
+    return Status::OK();
+  }
+
+  Status Visit(const UnionType& type) {
+    const auto& child_ids = type.child_ids();
+    const auto& type_codes_map = type.type_codes();
+
+    const int8_t* type_codes = data.GetValues<int8_t>(1);
+
+    for (int64_t i = 0; i < data.length; ++i) {
+      // Note that union arrays never have top-level nulls
       const int32_t code = type_codes[i];
       if (code < 0 || child_ids[code] == UnionType::kInvalidChildId) {
         return Status::Invalid("Union value at position ", i, " has invalid type id ",
                                code);
       }
     }
```

Comment:
Should we also check that no type code is larger than the largest type code in type_codes_map?
```suggestion
    const auto& type_codes_map = type.type_codes();
    if (type_codes_map.empty()) return Status::OK();

    int8_t max_type_code = *std::max_element(type_codes_map.begin(), type_codes_map.end());
    const int8_t* type_codes = data.GetValues<int8_t>(1);

    for (int64_t i = 0; i < data.length; ++i) {
      // Note that union arrays never have top-level nulls
      const int32_t code = type_codes[i];
      if (code < 0 || code > max_type_code || child_ids[code] == UnionType::kInvalidChildId) {
        return Status::Invalid("Union value at position ", i, " has invalid type id ",
                               code);
      }
    }
```

Reply:
`child_ids` always has 128 entries, so this shouldn't be necessary.
https://github.com/apache/arrow/blob/master/cpp/src/arrow/type.cc#L582